src/python/pants/backend/python/tasks/python_bundle.py

```diff
@@ -0,0 +1,95 @@
+    cls.register_common_options(register)
+
+  @classmethod
+  def product_types(cls):
+    return [cls._DEPLOYABLE_ARCHIVES]
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(PythonBundle, cls).prepare(options, round_manager)
+    round_manager.require_data(cls._PEX_ARCHIVES)
+
+  @staticmethod
+  def _get_archive_path(vt, archive_format):
+    ext = archive.archive_extensions.get(archive_format, archive_format)
+    filename = '{}.{}'.format(vt.target.id, ext)
+    return os.path.join(vt.results_dir, filename) if archive_format else ''
+
+  @property
+  def create_target_dirs(self):
+    return True
+
+  def __init__(self, *args, **kwargs):
+    super(PythonBundle, self).__init__(*args, **kwargs)
+    self._outdir = self.get_options().pants_distdir
+
+  def execute(self):
+    targets_to_bundle = self.context.targets(PythonApp.is_python_app)
+
+    with self.invalidated(targets_to_bundle, invalidate_dependents=True) as invalidation_check:
+      bundle_archive_product = self.context.products.get(self._DEPLOYABLE_ARCHIVES)
+
+      for vt in invalidation_check.all_vts:
+        bundle_dir = self.get_bundle_dir(vt.target.id, vt.results_dir)
+        archive_format = self.resolved_option(self.get_options(), vt.target, 'archive')
+        archiver = archive.archiver(archive_format) if archive_format else None
+        archive_path = self._get_archive_path(vt, archive_format)
+
+        if not vt.valid:  # Only recreate the bundle/archive if it's changed
+          self._bundle(vt.target, bundle_dir)
+          if archiver:
+            archiver.create(bundle_dir, vt.results_dir, vt.target.id)
+            self.context.log.info(
+              'created archive {}'.format(os.path.relpath(archive_path, get_buildroot())))
+
+        if archiver:
+          bundle_archive_product.add(
+            vt.target, os.path.dirname(archive_path)).append(os.path.basename(archive_path))
+
+        if vt.target in self.context.target_roots:  # Always publish bundle/archive in dist
+          self.publish_results(self.get_options().pants_distdir,
+                               False,
+                               vt,
+                               bundle_dir,
+                               archive_path,
+                               vt.target.id,
+                               archive_format)
+
+  def _bundle(self, target, bundle_dir):
+    self.context.log.debug('creating {}'.format(os.path.relpath(bundle_dir, get_buildroot())))
+    safe_mkdir(bundle_dir, clean=True)
+    binary_path = self._get_binary_path(target)
+    os.symlink(binary_path, os.path.join(bundle_dir, os.path.basename(binary_path)))
+    self.symlink_bundles(target, bundle_dir)
+
+  def _get_binary_path(self, target):
+    pex_archives = self.context.products.get(self._PEX_ARCHIVES)
+    for basedir, filenames in pex_archives.get(target.binary).items():
+      for filename in filenames:
+        return os.path.join(basedir, filename)
```

Comment:
AFAICT `PythonBinary` targets always have exactly one pex product, so I _think_ this is safe. I'll add an assert and update the comments.

That said, long-term I would love to have a generic bundle task that can produce a deployable archive with multiple binaries of potentially different languages. We already have a use-case where we bundle a `jvm_app`, then (with an external script) add a `pex` to the archive before deploying.

Reply:
@traviscrawford do you think the deployable archive could have a language-agnostic entry point (...doesn't seem ridiculous), or do you mean the output would remain some existing archive format (as in, we only produce language-specific archives like pex, but we can nest in other bundles)?

Reply:
Imagine you want to create an `archive.tar.gz` that looks like:

```
./bin/foo.pex
./bin/bar.pex
./bin/baz.jar
./etc/foo.ini
./etc/bar.cfg
./etc/baz.xml
```

Where it has multiple binaries, multiple "extra files", etc. That feels like something Pants could bundle up for you, rather than having the script multiple invocations of Pants to make all the artifacts, and to archive things up at the end.

This "fancy bundling" is not included in this change, but directionally I think the additional flexibility would be really valuable for users.